File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/external/tax/usecases/GetTaxRate.java
package com.mycompany.myshop.testkit.dsl.core.usecase.external.tax.usecases;

import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.tax.usecases.base.BaseTaxUseCase;
import com.mycompany.myshop.testkit.driver.port.external.tax.TaxDriver;
import com.mycompany.myshop.testkit.driver.port.external.tax.dtos.GetTaxResponse;
import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;

public class GetTaxRate extends BaseTaxUseCase<GetTaxResponse, GetTaxVerification> {
    private String countryValueOrAlias;

    public GetTaxRate(TaxDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public GetTaxRate country(String countryValueOrAlias) {
        this.countryValueOrAlias = countryValueOrAlias;
        return this;
    }

    @Override
    public UseCaseResult<GetTaxResponse, GetTaxVerification> execute() {
        var country = context.getParamValueOrLiteral(countryValueOrAlias);

        var result = driver.getTaxRate(country);

        return new UseCaseResult<>(result.mapError(e -> SystemError.of(e.getMessage())), context, GetTaxVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/tax/usecases/GetTaxRate.ts
import type { TaxDriver } from '../../../../../../driver/port/external/tax/tax-driver.js';
import type { GetCountryRequest } from '../../../../../../driver/port/external/tax/dtos/GetCountryRequest.js';
import type { GetTaxResponse } from '../../../../../../driver/port/external/tax/dtos/GetTaxResponse.js';
import { success, failure, type Result } from '../../../../../../common/result.js';
import type { SystemError } from '../../../../../../driver/port/dtos/errors/SystemError.js';
import { UseCaseResult } from '../../../../shared/use-case-result.js';
import type { UseCaseContext } from '../../../../shared/use-case-context.js';
import { BaseTaxUseCase } from './base/BaseTaxUseCase.js';
import { GetTaxVerification } from './GetTaxVerification.js';

export class GetTaxRate extends BaseTaxUseCase<GetTaxResponse, GetTaxVerification> {
  private _countryValueOrAlias?: string;

  constructor(driver: TaxDriver, context: UseCaseContext) {
    super(driver, context);
  }

  country(countryValueOrAlias: string): this {
    this._countryValueOrAlias = countryValueOrAlias;
    return this;
  }

  async execute(): Promise<UseCaseResult<GetTaxResponse, GetTaxVerification>> {
    const country = this.context.getParamValueOrLiteral(this._countryValueOrAlias);

    const request: GetCountryRequest = {
      country: country ?? '',
    };

    const driverResult = await this.driver.getTaxRate(request);
    const result: Result<GetTaxResponse, SystemError> = driverResult.success
      ? success(driverResult.value)
      : failure({ message: driverResult.error.message ?? 'Tax error', fieldErrors: [] });

    return new UseCaseResult(
      result,
      this.context,
      (response, ctx) => new GetTaxVerification(response, ctx),
    );
  }
}
